File: Bilibili/iPhone/Categories/CategoriesViewController.m
//
//  CategoriesViewController.m
//  Bilibili
//
//  Created by LunarEclipse on 16/9/2.
//  Copyright © 2016年 LunarEclipse. All rights reserved.
//

#import "CategoriesViewController.h"
#import "LiveHomeHeader.h"
#import "CategoryCell.h"
#import "FlowCollectionView.h"

#define ID @"CategoryCell"

@import YYKit;

@interface CategoriesViewController ()

@property (nonatomic, strong) UICollectionView * collectionView;

@property (nonatomic, copy) NSArray * images;
@property (nonatomic, copy) NSArray * names;

@end

@implementation CategoriesViewController

#pragma mark Override

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.titleName = @"分区";
    
    _images = @[@"home_region_icon_live", @"home_region_icon_13", @"home_region_icon_1", @"home_region_icon_3",
                @"home_region_icon_129", @"home_region_icon_4", @"home_region_icon_36", @"home_region_icon_160",
                @"home_region_icon_119", @"home_region_icon_fashion", @"home_region_icon_165", @"home_region_icon_155",
                @"home_region_icon_23", @"home_region_icon_5"];
    
    _names = @[@"直播", @"番剧", @"动画", @"音乐", @"舞蹈", @"游戏", @"科技", @"生活", @"鬼畜",
               @"时尚", @"广告", @"娱乐", @"电影", @"电视剧", @"游戏中心"];
    
    //Initialize CollectionView
    UICollectionViewFlowLayout * layout = [UICollectionViewFlowLayout new];
    layout.minimumLineSpacing = 0;
    layout.minimumInteritemSpacing = 8;
    
    _collectionView = [[UICollectionView alloc] initWithFrame:self.contentView.bounds collectionViewLayout:layout];
    _collectionView.showsVerticalScrollIndicator = NO;
    _collectionView.backgroundColor = [UIColor colorWithWhite:0.95 alpha:1.000];
    _collectionView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _collectionView.contentInset = UIEdgeInsetsMake(20, 20, 0, 20);
    _collectionView.dataSource = self;
    _collectionView.delegate = self;
    [_collectionView registerNib:[UINib nibWithNibName:@"CategoryCell" bundle:nil] forCellWithReuseIdentifier:ID];
    [self.contentView addSubview:_collectionView];
    
    CGFloat length = floorf(kScreenWidth / 3) - ceilf(layout.minimumInteritemSpacing * 2 / 3) - ceilf((_collectionView.contentInset.left + _collectionView.contentInset.right) / 3);
    layout.itemSize = CGSizeMake(length, length);
}

#pragma mark Delegate

/* UICollectionViewDataSource */

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
{
    return 1;
}

-(NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    return MIN(_images.count, _names.count);
}

-(UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    CategoryCell * cell = [_collectionView dequeueReusableCellWithReuseIdentifier:ID forIndexPath:indexPath];
    cell.iconImageView.image = [UIImage imageNamed:_images[indexPath.row]];
    cell.nameLabel.text = _names[indexPath.row];
    return cell;
}

@end

File: Bilibili/iPhone/Categories/CategoriesViewController.js
require('UICollectionViewFlowLayout,UICollectionView,UIScreen,UIColor,CategoryCell,UIImage,NSIndexPath');
require('JPEngine').defineStruct({
                                 "name": "UIEdgeInsets",
                                 "types": "FFFF",
                                 "keys": ["top", "left", "bottom", "right"]
})

defineClass('CategoriesViewController', {
            viewDidLoad: function() {
            self.super().viewDidLoad();
            
            _images = ["home_region_icon_live", "home_region_icon_13", "home_region_icon_1", "home_region_icon_3",
                       "home_region_icon_129", "home_region_icon_4", "home_region_icon_36", "home_region_icon_life",
                       "home_region_icon_119", "home_region_icon_fashion", "home_region_gamecenter", "home_region_icon_entertainment",
                       "home_region_icon_23", "home_region_icon_5"
                       ];
            
            var width = UIScreen.mainScreen().bounds().width;
            
            //Initialize CollectionView
            var layout = UICollectionViewFlowLayout.new();
            layout.setMinimumLineSpacing(16);
            layout.setMinimumInteritemSpacing(24);
            layout.setItemSize({width: width / 3 - 24 * 2, height:width / 3 - 24 * 2});
            
            _collectionView = UICollectionView.alloc().initWithFrame_collectionViewLayout(self.view().bounds(), layout);
            _collectionView.setScrollEnabled(NO);
            _collectionView.setBackgroundColor(UIColor.colorWithWhite_alpha(0.95, 1.000));
            _collectionView.setAutoresizingMask(1 << 1 | 1 << 4);  //FlexibleWidth | FlexibleHeight
            _collectionView.setContentInset({top: 20, left: 20, bottom: 20, right: 20});
            _collectionView.setDataSource(self);
            _collectionView.setDelegate(self);
            _collectionView.registerClass_forCellWithReuseIdentifier(CategoryCell.class(), "CategoryCell");
            self.contentView().addSubview(_collectionView);
            },
            numberOfSectionsInCollectionView: function(collectionView) {
                return 1;
            },
            
            collectionView_numberOfItemsInSection: function(collectionView, section) {
                return _images.length;
            },
            
            collectionView_cellForItemAtIndexPath: function(collectionView, indexPath) {
                var cell = _collectionView.dequeueReusableCellWithReuseIdentifier_forIndexPath("CategoryCell", indexPath);
                cell.imageView().setImage(UIImage.imageNamed(_images[indexPath.row()]));
                return cell;
            },
});